File: api/app/modules/queues/solver.py
from __future__ import annotations

import logging
import math
from typing import Any

from app.modules.queues.models import QueuesRequest
from app.schemas.common import SolveStatus
from app.schemas.result import GraphXY, ModuleResult, NamedTable, SolutionBlock

logger = logging.getLogger(__name__)

INF = float("inf")


def solve(req: QueuesRequest) -> ModuleResult:
    lam = req.lambda_
    mu = req.mu
    warnings: list[str] = []

    metrics, pn = _dispatch(req, lam, mu, warnings)

    tables: list[NamedTable] = []
    graph = None
    if pn is not None and req.include_pn:
        tables.append(
            NamedTable(
                name="Pn",
                columns=["n", "Pn"],
                rows=[[i, float(p)] for i, p in enumerate(pn)],
            )
        )
        graph = GraphXY(
            type="xy",
            series=[{"name": "Pn", "x": list(range(len(pn))), "y": [float(p) for p in pn]}],
            x_label="n (clientes en el sistema)",
            y_label="Probabilidad Pn",
            title="Distribución de probabilidad del número de clientes",
            subtitle=f"Modelo {req.model}",
            kind="bar",
        )

    _apply_costs(req, metrics, warnings)

    variables = {
        k: float(v)
        for k, v in metrics.items()
        if isinstance(v, (int, float)) and math.isfinite(float(v))
    }

    if req.optimize_s:
        cost_table = _optimize_servers(req, lam, mu, warnings)
        if cost_table is not None:
            tables.append(cost_table)

    result = ModuleResult(
        module="queues",
        status=SolveStatus.ok,
        solution=SolutionBlock(variables=variables, metrics=metrics),
        iterations=None,
        sensitivity=None,
        graph=graph,
        tables=tables or None,
        warnings=warnings,
    )
    logger.info("module=%s status=%s", result.module, result.status.value)
    return result


def _dispatch(
    req: QueuesRequest, lam: float, mu: float, warnings: list[str]
) -> tuple[dict[str, float], list[float] | None]:
    if req.model == "M/M/1":
        return _mm1(lam, mu, req.include_pn, warnings)
    if req.model == "M/M/s":
        s = int(req.s)  # type: ignore[arg-type]
        return _mms(lam, mu, s, req.include_pn, warnings)
    if req.model == "M/M/1/K":
        K = int(req.K)  # type: ignore[arg-type]
        return _mm1k(lam, mu, K, warnings)
    if req.model == "M/M/s/N":
        s = int(req.s)  # type: ignore[arg-type]
        N = int(req.N)  # type: ignore[arg-type]
        return _mmsn(lam, mu, s, N, warnings)
    if req.model == "M/M/s/K":
        s = int(req.s)  # type: ignore[arg-type]
        K = int(req.K)  # type: ignore[arg-type]
        return _mmsk(lam, mu, s, K, warnings)
    if req.model == "M/G/1":
        sigma = float(req.service_std_dev)  # type: ignore[arg-type]
        return _mg1(lam, mu, sigma, warnings)
    # M/D/1
    return _md1(lam, mu, warnings)


def _mm1(
    lam: float, mu: float, include_pn: bool, warnings: list[str]
) -> tuple[dict[str, float], list[float] | None]:
    rho = lam / mu
    if rho >= 1:
        warnings.append(
            "rho >= 1: el sistema M/M/1 es inestable (la cola crece sin límite); "
            "L, Lq, W y Wq tienden a infinito"
        )
        return {"L": INF, "Lq": INF, "W": INF, "Wq": INF, "rho": rho, "P0": 0.0}, None
    p0 = 1.0 - rho
    lq = rho**2 / (1.0 - rho)
    L = rho / (1.0 - rho)
    W = 1.0 / (mu - lam)
    wq = lam / (mu * (mu - lam))
    pn = [p0 * (rho**n) for n in range(0, 21)] if include_pn else None
    return {"L": L, "Lq": lq, "W": W, "Wq": wq, "rho": rho, "P0": p0}, pn


def _mms(
    lam: float, mu: float, s: int, include_pn: bool, warnings: list[str]
) -> tuple[dict[str, float], list[float] | None]:
    rho = lam / (s * mu)
    if rho >= 1:
        warnings.append(
            f"rho >= 1: el sistema M/M/{s} es inestable (lambda >= s*mu); "
            "L, Lq, W y Wq tienden a infinito"
        )
        return {"L": INF, "Lq": INF, "W": INF, "Wq": INF, "rho": rho, "P0": 0.0}, None
    sum_term = sum((lam / mu) ** n / math.factorial(n) for n in range(s))
    last = (lam / mu) ** s / (math.factorial(s) * (1.0 - rho))
    p0 = 1.0 / (sum_term + last)
    lq = p0 * ((lam / mu) ** s) * rho / (math.factorial(s) * (1.0 - rho) ** 2)
    L = lq + lam / mu
    wq = lq / lam
    W = wq + 1.0 / mu
    pn = None
    if include_pn:
        pn = []
        for n in range(0, max(s + 15, 21)):
            if n < s:
                pn.append(p0 * (lam / mu) ** n / math.factorial(n))
            else:
                pn.append(p0 * (lam / mu) ** n / (math.factorial(s) * s ** (n - s)))
    return {"L": L, "Lq": lq, "W": W, "Wq": wq, "rho": rho, "P0": p0}, pn


def _mm1k(lam: float, mu: float, K: int, warnings: list[str]) -> tuple[dict[str, float], list[float]]:
    r = lam / mu
    if abs(r - 1.0) < 1e-12:
        p0 = 1.0 / (K + 1)
        pn = [p0] * (K + 1)
    else:
        p0 = (1.0 - r) / (1.0 - r ** (K + 1))
        pn = [p0 * (r**n) for n in range(K + 1)]
    L = sum(n * pn[n] for n in range(K + 1))
    lam_eff = lam * (1.0 - pn[K])
    lq = L - (lam_eff / mu)
    W = L / lam_eff if lam_eff > 1e-12 else INF
    wq = lq / lam_eff if lam_eff > 1e-12 else INF
    rho = 1.0 - p0
    warnings.append("rho reportado como 1-P0 (utilización del servidor) para capacidad finita")
    return {
        "L": L,
        "Lq": lq,
        "W": W,
        "Wq": wq,
        "rho": rho,
        "P0": p0,
        "lambda_eff": lam_eff,
    }, pn


def _mmsn(
    lam: float, mu: float, s: int, N: int, warnings: list[str]
) -> tuple[dict[str, float], list[float]]:
    pn_unnorm = [1.0]
    for n in range(1, N + 1):
        if n <= s:
            pn_unnorm.append(pn_unnorm[-1] * (N - n + 1) * lam / (n * mu))
        else:
            pn_unnorm.append(pn_unnorm[-1] * (N - n + 1) * lam / (s * mu))
    total = sum(pn_unnorm)
    pn = [x / total for x in pn_unnorm]
    p0 = pn[0]
    L = sum(n * pn[n] for n in range(N + 1))
    lam_eff = lam * sum((N - n) * pn[n] for n in range(N + 1))
    busy = sum(min(n, s) * pn[n] for n in range(N + 1))
    lq = L - busy
    W = L / lam_eff if lam_eff > 1e-12 else INF
    wq = lq / lam_eff if lam_eff > 1e-12 else INF
    rho = busy / s
    if N <= s:
        warnings.append("N <= s: la cola nunca se forma en esta población finita (Lq = 0)")
    return {
        "L": L,
        "Lq": lq,
        "W": W,
        "Wq": wq,
        "rho": rho,
        "P0": p0,
        "lambda_eff": lam_eff,
    }, pn


def _mmsk(lam: float, mu: float, s: int, K: int, warnings: list[str]) -> tuple[dict[str, float], list[float]]:
    if K < s:
        raise ValueError("K debe ser mayor o igual que s en M/M/s/K")
    r = lam / mu
    pn_unnorm = [1.0]
    for n in range(1, K + 1):
        if n <= s:
            pn_unnorm.append(pn_unnorm[-1] * r / n)
        else:
            pn_unnorm.append(pn_unnorm[-1] * r / s)
    total = sum(pn_unnorm)
    pn = [x / total for x in pn_unnorm]
    p0 = pn[0]
    L = sum(n * pn[n] for n in range(K + 1))
    lam_eff = lam * (1.0 - pn[K])
    busy = lam_eff / mu
    lq = L - busy
    W = L / lam_eff if lam_eff > 1e-12 else INF
    wq = lq / lam_eff if lam_eff > 1e-12 else INF
    rho = busy / s
    return {
        "L": L,
        "Lq": lq,
        "W": W,
        "Wq": wq,
        "rho": rho,
        "P0": p0,
        "lambda_eff": lam_eff,
    }, pn


def _mg1(lam: float, mu: float, sigma: float, warnings: list[str]) -> tuple[dict[str, float], None]:
    rho = lam / mu
    if rho >= 1:
        warnings.append(
            "rho >= 1: el sistema M/G/1 es inestable; L, Lq, W y Wq tienden a infinito"
        )
        return {"L": INF, "Lq": INF, "W": INF, "Wq": INF, "rho": rho, "P0": 0.0}, None
    lq = (lam**2 * sigma**2 + rho**2) / (2.0 * (1.0 - rho))
    L = lq + rho
    wq = lq / lam
    W = wq + 1.0 / mu
    p0 = 1.0 - rho
    warnings.append(
        "M/G/1 usa la fórmula de Pollaczek-Khinchine; no existe distribución Pn de forma cerrada"
    )
    return {"L": L, "Lq": lq, "W": W, "Wq": wq, "rho": rho, "P0": p0}, None


def _md1(lam: float, mu: float, warnings: list[str]) -> tuple[dict[str, float], None]:
    metrics, _ = _mg1(lam, mu, 0.0, warnings)
    return metrics, None


def _apply_costs(req: QueuesRequest, metrics: dict[str, float], warnings: list[str]) -> None:
    if req.cost_waiting_per_unit_time is None and req.cost_server_per_unit_time is None:
        return
    if metrics.get("L") == INF:
        warnings.append("No se calculan costos: el sistema es inestable (L = infinito)")
        return
    s_used = req.s if req.s is not None else 1
    cw = req.cost_waiting_per_unit_time or 0.0
    cs = req.cost_server_per_unit_time or 0.0
    cost_waiting = cw * metrics["L"]
    cost_server = cs * s_used
    metrics["cost_waiting"] = float(cost_waiting)
    metrics["cost_server"] = float(cost_server)
    metrics["cost_total"] = float(cost_waiting + cost_server)


def _optimize_servers(
    req: QueuesRequest, lam: float, mu: float, warnings: list[str]
) -> NamedTable | None:
    if req.model not in ("M/M/s", "M/M/s/K", "M/M/s/N"):
        warnings.append("optimize_s solo aplica a modelos multi-servidor (M/M/s, M/M/s/K, M/M/s/N)")
        return None
    if req.cost_waiting_per_unit_time is None or req.cost_server_per_unit_time is None:
        warnings.append(
            "optimize_s requiere cost_waiting_per_unit_time y cost_server_per_unit_time"
        )
        return None

    cw = req.cost_waiting_per_unit_time
    cs = req.cost_server_per_unit_time
    s_min = max(1, math.ceil(lam / mu) if req.model != "M/M/s/N" else 1)
    s_max = req.s_max or max(s_min + 10, (req.s or s_min) + 10)

    rows: list[list[Any]] = []
    best_s = None
    best_cost = INF
    local_warnings: list[str] = []
    for s_try in range(1, s_max + 1):
        try:
            if req.model == "M/M/s":
                m, _ = _mms(lam, mu, s_try, False, local_warnings)
            elif req.model == "M/M/s/K":
                K = int(req.K)  # type: ignore[arg-type]
                if K < s_try:
                    continue
                m, _ = _mmsk(lam, mu, s_try, K, local_warnings)
            else:
                N = int(req.N)  # type: ignore[arg-type]
                if s_try > N:
                    continue
                m, _ = _mmsn(lam, mu, s_try, N, local_warnings)
        except ValueError:
            continue
        if m.get("L") == INF:
            continue
        cost_waiting = cw * m["L"]
        cost_server = cs * s_try
        cost_total = cost_waiting + cost_server
        rows.append([s_try, m["L"], m["Lq"], m["W"], m["Wq"], m["rho"], cost_waiting, cost_server, cost_total, False])
        if cost_total < best_cost:
            best_cost = cost_total
            best_s = s_try

    if best_s is None:
        warnings.append("No fue posible optimizar s: ningún valor evaluado resultó estable")
        return None

    for row in rows:
        row[-1] = row[0] == best_s

    warnings.append(f"Número óptimo de servidores recomendado: s = {best_s} (costo total = {best_cost:.4g})")

    return NamedTable(
        name="cost_by_s",
        columns=[
            "servidores",
            "L",
            "Lq",
            "W",
            "Wq",
            "rho",
            "costo_espera",
            "costo_servidor",
            "costo_total",
            "óptimo",
        ],
        rows=rows,
    )

File: worker/src/modules/queues/solver.ts
import { SolverError } from "../../errors";
import { LIMITS, assertLimit } from "../../limits";
import type { ModuleResult, NamedTable } from "../../schema";
import { okResult } from "../../schema";

const INF = Number.POSITIVE_INFINITY;

export type QueueModel = "M/M/1" | "M/M/s" | "M/M/1/K" | "M/M/s/N" | "M/M/s/K" | "M/G/1" | "M/D/1";

export type QueuesRequest = {
  model: QueueModel;
  lambda: number;
  mu: number;
  s?: number | null;
  K?: number | null;
  N?: number | null;
  service_std_dev?: number | null;
  include_pn?: boolean;
  cost_waiting_per_unit_time?: number | null;
  cost_server_per_unit_time?: number | null;
  optimize_s?: boolean;
  s_max?: number | null;
};

function asRecord(body: unknown): Record<string, unknown> {
  if (!body || typeof body !== "object" || Array.isArray(body)) {
    throw new SolverError("payload debe ser un objeto JSON");
  }
  return body as Record<string, unknown>;
}

export function parseQueuesRequest(body: unknown): QueuesRequest {
  const o = asRecord(body);
  const model = o.model as QueueModel;
  const allowed: QueueModel[] = ["M/M/1", "M/M/s", "M/M/1/K", "M/M/s/N", "M/M/s/K", "M/G/1", "M/D/1"];
  if (!allowed.includes(model)) throw new SolverError("modelo de colas no soportado");
  const lam = Number(o.lambda);
  const mu = Number(o.mu);
  if (!(lam > 0) || !(mu > 0)) throw new SolverError("lambda y mu deben ser > 0");
  const req: QueuesRequest = {
    model,
    lambda: lam,
    mu,
    s: o.s == null ? null : Number(o.s),
    K: o.K == null ? null : Number(o.K),
    N: o.N == null ? null : Number(o.N),
    service_std_dev: o.service_std_dev == null ? null : Number(o.service_std_dev),
    include_pn: o.include_pn !== false,
    cost_waiting_per_unit_time: o.cost_waiting_per_unit_time == null ? null : Number(o.cost_waiting_per_unit_time),
    cost_server_per_unit_time: o.cost_server_per_unit_time == null ? null : Number(o.cost_server_per_unit_time),
    optimize_s: Boolean(o.optimize_s),
    s_max: o.s_max == null ? null : Number(o.s_max),
  };
  if (["M/M/s", "M/M/s/N", "M/M/s/K"].includes(model) && req.s == null) {
    throw new SolverError("s is required for multi-server models");
  }
  if (model === "M/M/1/K" && req.K == null) throw new SolverError("K is required for M/M/1/K");
  if (model === "M/M/s/K" && req.K == null) throw new SolverError("K is required for M/M/s/K");
  if (model === "M/M/s/N" && req.N == null) throw new SolverError("N is required for M/M/s/N");
  if (model === "M/G/1" && req.service_std_dev == null) {
    throw new SolverError("service_std_dev is required for M/G/1");
  }
  if (req.s != null) assertLimit(req.s <= LIMITS.queuesSMax, `s no puede superar ${LIMITS.queuesSMax}`);
  if (req.K != null) assertLimit(req.K <= LIMITS.queuesK, `K no puede superar ${LIMITS.queuesK}`);
  if (req.N != null) assertLimit(req.N <= LIMITS.queuesN, `N no puede superar ${LIMITS.queuesN}`);
  if (req.s_max != null) assertLimit(req.s_max <= LIMITS.queuesSMax, `s_max no puede superar ${LIMITS.queuesSMax}`);
  return req;
}

function factorial(n: number): number {
  let x = 1;
  for (let i = 2; i <= n; i++) x *= i;
  return x;
}

export function solve(body: unknown): ModuleResult {
  const req = parseQueuesRequest(body);
  const lam = req.lambda;
  const mu = req.mu;
  const warnings: string[] = [];
  const { metrics, pn } = dispatch(req, lam, mu, warnings);

  const tables: NamedTable[] = [];
  let graph = null;
  if (pn && req.include_pn) {
    tables.push({
      name: "Pn",
      columns: ["n", "Pn"],
      rows: pn.map((p, i) => [i, p]),
    });
    graph = {
      type: "xy" as const,
      series: [{ name: "Pn", x: pn.map((_, i) => i), y: pn.map((p) => p) }],
      x_label: "n (clientes en el sistema)",
      y_label: "Probabilidad Pn",
      title: "Distribución de probabilidad del número de clientes",
      subtitle: `Modelo ${req.model}`,
      kind: "bar",
    };
  }

  applyCosts(req, metrics, warnings);
  const variables: Record<string, number> = {};
  for (const [k, v] of Object.entries(metrics)) {
    if (typeof v === "number" && Number.isFinite(v)) variables[k] = v;
  }

  if (req.optimize_s) {
    const costTable = optimizeServers(req, lam, mu, warnings);
    if (costTable) tables.push(costTable);
  }

  return okResult("queues", {
    status: "ok",
    variables,
    metrics,
    graph,
    tables: tables.length ? tables : null,
    warnings,
  });
}

function dispatch(
  req: QueuesRequest,
  lam: number,
  mu: number,
  warnings: string[],
): { metrics: Record<string, number>; pn: number[] | null } {
  if (req.model === "M/M/1") return mm1(lam, mu, req.include_pn !== false, warnings);
  if (req.model === "M/M/s") return mms(lam, mu, Number(req.s), req.include_pn !== false, warnings);
  if (req.model === "M/M/1/K") return mm1k(lam, mu, Number(req.K), warnings);
  if (req.model === "M/M/s/N") return mmsn(lam, mu, Number(req.s), Number(req.N), warnings);
  if (req.model === "M/M/s/K") return mmsk(lam, mu, Number(req.s), Number(req.K), warnings);
  if (req.model === "M/G/1") return mg1(lam, mu, Number(req.service_std_dev), warnings);
  return md1(lam, mu, warnings);
}

function mm1(lam: number, mu: number, includePn: boolean, warnings: string[]) {
  const rho = lam / mu;
  if (rho >= 1) {
    warnings.push(
      "rho >= 1: el sistema M/M/1 es inestable (la cola crece sin límite); L, Lq, W y Wq tienden a infinito",
    );
    return { metrics: { L: INF, Lq: INF, W: INF, Wq: INF, rho, P0: 0 }, pn: null };
  }
  const p0 = 1 - rho;
  const lq = rho ** 2 / (1 - rho);
  const L = rho / (1 - rho);
  const W = 1 / (mu - lam);
  const wq = lam / (mu * (mu - lam));
  const pn = includePn ? Array.from({ length: 21 }, (_, n) => p0 * rho ** n) : null;
  return { metrics: { L, Lq: lq, W, Wq: wq, rho, P0: p0 }, pn };
}

function mms(lam: number, mu: number, s: number, includePn: boolean, warnings: string[]) {
  const rho = lam / (s * mu);
  if (rho >= 1) {
    warnings.push(
      `rho >= 1: el sistema M/M/${s} es inestable (lambda >= s*mu); L, Lq, W y Wq tienden a infinito`,
    );
    return { metrics: { L: INF, Lq: INF, W: INF, Wq: INF, rho, P0: 0 }, pn: null };
  }
  let sumTerm = 0;
  for (let n = 0; n < s; n++) sumTerm += (lam / mu) ** n / factorial(n);
  const last = (lam / mu) ** s / (factorial(s) * (1 - rho));
  const p0 = 1 / (sumTerm + last);
  const lq = (p0 * (lam / mu) ** s * rho) / (factorial(s) * (1 - rho) ** 2);
  const L = lq + lam / mu;
  const wq = lq / lam;
  const W = wq + 1 / mu;
  let pn: number[] | null = null;
  if (includePn) {
    const nMax = Math.max(s + 15, 21);
    pn = [];
    for (let n = 0; n < nMax; n++) {
      if (n < s) pn.push((p0 * (lam / mu) ** n) / factorial(n));
      else pn.push((p0 * (lam / mu) ** n) / (factorial(s) * s ** (n - s)));
    }
  }
  return { metrics: { L, Lq: lq, W, Wq: wq, rho, P0: p0 }, pn };
}

function mm1k(lam: number, mu: number, K: number, warnings: string[]) {
  const r = lam / mu;
  let p0: number;
  let pn: number[];
  if (Math.abs(r - 1) < 1e-12) {
    p0 = 1 / (K + 1);
    pn = Array(K + 1).fill(p0);
  } else {
    p0 = (1 - r) / (1 - r ** (K + 1));
    pn = Array.from({ length: K + 1 }, (_, n) => p0 * r ** n);
  }
  const L = pn.reduce((s, p, n) => s + n * p, 0);
  const lamEff = lam * (1 - pn[K]);
  const lq = L - lamEff / mu;
  const W = lamEff > 1e-12 ? L / lamEff : INF;
  const wq = lamEff > 1e-12 ? lq / lamEff : INF;
  const rho = 1 - p0;
  warnings.push("rho reportado como 1-P0 (utilización del servidor) para capacidad finita");
  return { metrics: { L, Lq: lq, W, Wq: wq, rho, P0: p0, lambda_eff: lamEff }, pn };
}

function mmsn(lam: number, mu: number, s: number, N: number, warnings: string[]) {
  const pnUnnorm = [1];
  for (let n = 1; n <= N; n++) {
    if (n <= s) pnUnnorm.push((pnUnnorm[n - 1] * (N - n + 1) * lam) / (n * mu));
    else pnUnnorm.push((pnUnnorm[n - 1] * (N - n + 1) * lam) / (s * mu));
  }
  const total = pnUnnorm.reduce((a, b) => a + b, 0);
  const pn = pnUnnorm.map((x) => x / total);
  const p0 = pn[0];
  const L = pn.reduce((s0, p, n) => s0 + n * p, 0);
  const lamEff = lam * pn.reduce((s0, p, n) => s0 + (N - n) * p, 0);
  const busy = pn.reduce((s0, p, n) => s0 + Math.min(n, s) * p, 0);
  const lq = L - busy;
  const W = lamEff > 1e-12 ? L / lamEff : INF;
  const wq = lamEff > 1e-12 ? lq / lamEff : INF;
  const rho = busy / s;
  if (N <= s) warnings.push("N <= s: la cola nunca se forma en esta población finita (Lq = 0)");
  return { metrics: { L, Lq: lq, W, Wq: wq, rho, P0: p0, lambda_eff: lamEff }, pn };
}

function mmsk(lam: number, mu: number, s: number, K: number, warnings: string[]) {
  if (K < s) throw new SolverError("K debe ser mayor o igual que s en M/M/s/K");
  const r = lam / mu;
  const pnUnnorm = [1];
  for (let n = 1; n <= K; n++) {
    if (n <= s) pnUnnorm.push((pnUnnorm[n - 1] * r) / n);
    else pnUnnorm.push((pnUnnorm[n - 1] * r) / s);
  }
  const total = pnUnnorm.reduce((a, b) => a + b, 0);
  const pn = pnUnnorm.map((x) => x / total);
  const p0 = pn[0];
  const L = pn.reduce((s0, p, n) => s0 + n * p, 0);
  const lamEff = lam * (1 - pn[K]);
  const busy = lamEff / mu;
  const lq = L - busy;
  const W = lamEff > 1e-12 ? L / lamEff : INF;
  const wq = lamEff > 1e-12 ? lq / lamEff : INF;
  const rho = busy / s;
  void warnings;
  return { metrics: { L, Lq: lq, W, Wq: wq, rho, P0: p0, lambda_eff: lamEff }, pn };
}

function mg1(lam: number, mu: number, sigma: number, warnings: string[]) {
  const rho = lam / mu;
  if (rho >= 1) {
    warnings.push("rho >= 1: el sistema M/G/1 es inestable; L, Lq, W y Wq tienden a infinito");
    return { metrics: { L: INF, Lq: INF, W: INF, Wq: INF, rho, P0: 0 }, pn: null };
  }
  const lq = (lam ** 2 * sigma ** 2 + rho ** 2) / (2 * (1 - rho));
  const L = lq + rho;
  const wq = lq / lam;
  const W = wq + 1 / mu;
  const p0 = 1 - rho;
  warnings.push("M/G/1 usa la fórmula de Pollaczek-Khinchine; no existe distribución Pn de forma cerrada");
  return { metrics: { L, Lq: lq, W, Wq: wq, rho, P0: p0 }, pn: null };
}

function md1(lam: number, mu: number, warnings: string[]) {
  return mg1(lam, mu, 0, warnings);
}

function applyCosts(req: QueuesRequest, metrics: Record<string, number>, warnings: string[]) {
  if (req.cost_waiting_per_unit_time == null && req.cost_server_per_unit_time == null) return;
  if (metrics.L === INF) {
    warnings.push("No se calculan costos: el sistema es inestable (L = infinito)");
    return;
  }
  const sUsed = req.s ?? 1;
  const cw = req.cost_waiting_per_unit_time ?? 0;
  const cs = req.cost_server_per_unit_time ?? 0;
  const costWaiting = cw * metrics.L;
  const costServer = cs * sUsed;
  metrics.cost_waiting = costWaiting;
  metrics.cost_server = costServer;
  metrics.cost_total = costWaiting + costServer;
}

function optimizeServers(
  req: QueuesRequest,
  lam: number,
  mu: number,
  warnings: string[],
): NamedTable | null {
  if (!["M/M/s", "M/M/s/K", "M/M/s/N"].includes(req.model)) {
    warnings.push("optimize_s solo aplica a modelos multi-servidor (M/M/s, M/M/s/K, M/M/s/N)");
    return null;
  }
  if (req.cost_waiting_per_unit_time == null || req.cost_server_per_unit_time == null) {
    warnings.push("optimize_s requiere cost_waiting_per_unit_time y cost_server_per_unit_time");
    return null;
  }
  const cw = req.cost_waiting_per_unit_time;
  const cs = req.cost_server_per_unit_time;
  const sMin = Math.max(1, req.model !== "M/M/s/N" ? Math.ceil(lam / mu) : 1);
  const sMax = req.s_max ?? Math.max(sMin + 10, (req.s ?? sMin) + 10);
  const rows: unknown[][] = [];
  let bestS: number | null = null;
  let bestCost = INF;
  const local: string[] = [];
  for (let sTry = 1; sTry <= sMax; sTry++) {
    let m: Record<string, number>;
    try {
      if (req.model === "M/M/s") m = mms(lam, mu, sTry, false, local).metrics;
      else if (req.model === "M/M/s/K") {
        const K = Number(req.K);
        if (K < sTry) continue;
        m = mmsk(lam, mu, sTry, K, local).metrics;
      } else {
        const N = Number(req.N);
        if (sTry > N) continue;
        m = mmsn(lam, mu, sTry, N, local).metrics;
      }
    } catch {
      continue;
    }
    if (m.L === INF) continue;
    const costWaiting = cw * m.L;
    const costServer = cs * sTry;
    const costTotal = costWaiting + costServer;
    rows.push([sTry, m.L, m.Lq, m.W, m.Wq, m.rho, costWaiting, costServer, costTotal, false]);
    if (costTotal < bestCost) {
      bestCost = costTotal;
      bestS = sTry;
    }
  }
  if (bestS == null) {
    warnings.push("No fue posible optimizar s: ningún valor evaluado resultó estable");
    return null;
  }
  for (const row of rows) row[row.length - 1] = row[0] === bestS;
  warnings.push(`Número óptimo de servidores recomendado: s = ${bestS} (costo total = ${bestCost.toPrecision(4)})`);
  return {
    name: "cost_by_s",
    columns: [
      "servidores",
      "L",
      "Lq",
      "W",
      "Wq",
      "rho",
      "costo_espera",
      "costo_servidor",
      "costo_total",
      "óptimo",
    ],
    rows,
  };
}
